### DISCUSSION TITLE
How to adding disturbance in creating model?

### DISCUSSION THREAD

**Discussion by NekoHanko** [do-mpc/do-mpc#412] [Q&A]
Hi do-mpc team, I want to creat control model according to follow expression:
![image](https://github.com/do-mpc/do-mpc/assets/117264024/552f26a9-4009-4dcd-b26d-d26ce3fe57bb)
where dk is outside disturbance, In doc page, I do not find how to add this disturbance, I wish you can help me about this question，Thanks！

**Comment by JoshuaAda** [comment]
Hi,
first of all thank you for using Do-MPC. I believe you are referring to the term B_d*d(k)? If you take a look at our [introduction example](https://www.do-mpc.com/en/latest/getting_started.html), you notice that it is possible to define parameters using model.set_variable() in the same way you define states, only with the tag '_p' instead. After doing this you will set an uninitialised parameter in your case d(k) in your model. In every step of your simulation loop you will have to set the parameter d(k) using the [set_p_fun](https://www.do-mpc.com/en/latest/api/do_mpc.controller.MPC.html#set-p-fun) in both MPC and Simulator, where you set the value of d(k) randomly as you desire. Of cause, this only applies if you want to use a nominal MPC. For a robust MPC that considers different scenarios for specific values of d(k), have a look at the function [set_uncertainty_values()](https://www.do-mpc.com/en/latest/api/do_mpc.controller.MPC.html#do_mpc.controller.MPC.set_uncertainty_values) and our background article to [robust MPC](https://www.do-mpc.com/en/latest/theory_mpc.html#robust-multi-stage-nmpc).
Hope this helps,
Your Do-MPC Team

**Comment by NekoHanko** [comment]
Fully understanding，thanks a lot ！在 2024年4月5日，03:28，JoshuaAda ***@***.***> 写道：﻿
Hi,
first of all thank you for using Do-MPC. I believe you are referring to the term B_d*d(k)? If you take a look at our introduction example, you notice that it is possible to define parameters using model.set_variable() in the same way you define states, only with the tag '_p' instead. After doing this you will set an uninitialised parameter in your case d(k) in your model. In every step of your simulation loop you will have to set the parameter d(k) using the set_p_fun in both MPC and Simulator, where you set the value of d(k) randomly as you desire. Of cause, this only applies if you want to use a nominal MPC. For a robust MPC that considers different scenarios for specific values of d(k), have a look at the function set_uncertainty_values() and our background article to robust MPC.
Hope this helps,
Your Do-MPC Team
